pymongo_auth_aws/auth.py

```diff
@@ -189,10 +238,15 @@ def step(self, server_payload):
 
     def _first_payload(self):
         """Return the first SASL payload."""
+        global _cached_credential
         # If a username and password are not provided, drivers MUST query
         # a link-local AWS address for temporary credentials.
         if self._credentials.username is None:
-            self._credentials = _aws_temp_credentials()
+            try:
+                self._credentials = _aws_temp_credentials()
+            except Exception:
+                _cached_credential = None
+                raise
```

Comment:
Perhaps it makes more sense to only cache the credential _after_ auth succeeds? Hmm I suppose this logic to reset the credential would still be needed though. What do you think about making the try/except/reset cred logic a decorator that we can add to _first_payload+_second_payload? Then we can centralize the logic in one place.